src/_pytest/main.py

```diff
@@ -572,29 +629,30 @@ def _collect(self, arg):
 
                 for x in self._collectfile(path):
                     key = (type(x), x.fspath)
-                    if key in self._node_cache:
-                        yield self._node_cache[key]
+                    if key in self._node_cache2:
+                        yield self._node_cache2[key]
                     else:
-                        self._node_cache[key] = x
+                        self._node_cache2[key] = x
                         yield x
         else:
             assert argpath.check(file=1)
 
-            if argpath in self._node_cache:
-                col = self._node_cache[argpath]
+            if argpath in self._node_cache1:
+                col = self._node_cache1[argpath]
             else:
                 collect_root = self._pkg_roots.get(argpath.dirname, self)
                 col = collect_root._collectfile(argpath, handle_dupes=False)
                 if col:
-                    self._node_cache[argpath] = col
+                    self._node_cache1[argpath] = col
             m = self.matchnodes(col, names)
             # If __init__.py was the only file requested, then the matched node will be
             # the corresponding Package, and the first yielded item will be the __init__
             # Module itself, so just use that. If this special case isn't taken, then all
             # the files in the package will be yielded.
             if argpath.basename == "__init__.py":
+                assert isinstance(m[0], nodes.Collector)
                 try:
-                    yield next(m[0].collect())
+                    yield next(iter(m[0].collect()))
```

Comment:
Needed to add `iter()` here because I made the official type of `collect()` be `Iterable` and not `Iterator` (which isn't the case for all implementations). Calling `iter()` on an iterator should just be idempotent, so I think it's safe.